geodata/models/common.py

```diff
@@ -50,6 +43,58 @@ def __str__(self):
         return '{} {} (type: {})'.format(self.id, self.name, type(self))
 
 
+class ChecksumFile(ModifiableEntry):
+    """A base class for tracking files.
+
+    Child class must implement a file field called ``file``! This ensures the
+    field is initialized with coorect options like uplaod location and
+    validators.
+
+    """
+
+    name = models.CharField(max_length=100, blank=True, null=True)
+    checksum = models.CharField(max_length=64, blank=True, null=True)
+    compute_checksum = models.BooleanField(
+        default=False
+    )  # a flag to recompute the checksum on save
+    validate_checksum = models.BooleanField(
+        default=False
+    )  # a flag to validate the checksum against the saved checksum
+    last_validation = models.BooleanField(default=False)
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        # TODO: is there a cleaner way to enforce child class has `file` field?
+        if not hasattr(self, 'file'):
+            raise AttributeError('Child class of `ChecksumFile` must have a `file` field.')
+        if not self.name:
+            self.name = self.file.name
+        # Must save the model with the file before accessing it for the checksum
+        super(ChecksumFile, self).save(*args, **kwargs)
+        # Checksum is additional step after saving everything else - simply update these fields.
+        if self.compute_checksum or self.validate_checksum:
+            previous = self.checksum
+            with _field_file_to_local_path(self.file) as file_path:
+                self.checksum = compute_checksum(file_path)
+            if self.validate_checksum:
+                self.last_validation = self.checksum == previous
+            # Reset the user flags
+            self.compute_checksum = False
+            self.validate_checksum = False
+            # Simple update save - not full save
+            super(ChecksumFile, self).save(
+                update_fields=[
+                    'checksum',
+                    'compute_checksum',
+                    'last_validation',
+                    'validate_checksum',
+                ]
+            )
```

Comment:
@manthey, do we want to run this checksum-ing as a task? I tried doing so, but I couldn't think of a clever way of accessing any given subclass model. Since I made this an abstract class and expect child classes to have a `file` field, there's no manager for this model and I'm not sure how to easily query for this model entry within the task.